src/common/storage/src/cache.rs

```diff
@@ -59,46 +117,87 @@ impl CachePolicy for FuseCachePolicy {
     ) -> BoxFuture<'static, Result<(RpRead, BytesReader)>> {
         let path = path.to_string();
         let cache_path = self.cache_path(&path, &args);
+
+        let enable_async = self.enable_async;
+        let threshold = self.threshold;
+        // Record a visit to cache_path.
+        let count = self.visit.visit(cache_path.clone());
+        let concurrency = self.concurrency.clone();
+
         Box::pin(async move {
             match cache.read(&cache_path, OpRead::default()).await {
-                Ok(v) => Ok(v),
-                Err(err) if err.kind() == ErrorKind::ObjectNotFound => {
-                    let (rp, mut r) = inner.read(&path, args.clone()).await?;
-
-                    let size = rp.clone().into_metadata().content_length();
-                    // If size < 8MiB, we can optimize by buffer in memory.
-                    // TODO: make this configurable.
-                    if size <= 8 * 1024 * 1024 {
-                        let mut bs = Vec::with_capacity(size as usize);
-                        r.read_to_end(&mut bs).await.map_err(|err| {
-                            Error::new(
-                                ErrorKind::Unexpected,
-                                "read from underlying storage service",
-                            )
-                            .set_source(err)
-                        })?;
-                        let bs = Bytes::from(bs);
+                Ok(v) => return Ok(v),
+                Err(err) => {
+                    // If error's kind is not object not found, we should return
+                    // inner read directly. the cache services chould be down.
+                    if err.kind() != ErrorKind::ObjectNotFound {
+                        return inner.read(&path, args).await;
+                    }
+                    // The path is not warm enough, we just go back to inner.
+                    if count < threshold {
+                        return inner.read(&path, args).await;
+                    }
+                }
+            };
+
+            // permit will be dropped after cache filled.
+            let _permit = match concurrency.try_acquire() {
+                Ok(permit) => permit,
+                Err(_) => return inner.read(&path, args).await,
+            };
+
+            // Start filling cache.
+            let (rp, mut r) = inner.read(&path, args.clone()).await?;
+            let size = rp.clone().into_metadata().content_length();
 
+            // If size is small enough, we can return into memory.
+            if size < 4 * 1024 * 1024 {
+                let mut bs = Vec::with_capacity(size as usize);
+                r.read_to_end(&mut bs).await.map_err(|err| {
+                    Error::new(ErrorKind::Unexpected, "read from underlying storage")
+                        .set_source(err)
+                })?;
+                let bs = Bytes::from(bs);
+
+                if enable_async {
+                    let moved_bs = bs.clone();
```

Comment:
How about the memory usage if enable_async enabled?

Reply:
No big different as we only have `concurrent` running async tasks for cache.